src/reseeding.rs

```diff
@@ -39,73 +40,127 @@ use {RngCore, SeedableRng, Error, ErrorKind};
+where R: BlockRngCore<u32> + SeedableRng,
+      Rsdr: RngCore;
 
-impl<R: RngCore + SeedableRng, Rsdr: RngCore> ReseedingRng<R, Rsdr> {
+impl<R, Rsdr> ReseedingRng<R, Rsdr>
+where R: BlockRngCore<u32> + SeedableRng,
+      Rsdr: RngCore
+{
     /// Create a new `ReseedingRng` with the given parameters.
     ///
     /// # Arguments
     ///
     /// * `rng`: the random number generator to use.
     /// * `threshold`: the number of generated bytes after which to reseed the RNG.
     /// * `reseeder`: the RNG to use for reseeding.
-    pub fn new(rng: R, threshold: u64, reseeder: Rsdr) -> ReseedingRng<R,Rsdr> {
+    pub fn new(rng: R, threshold: u64, reseeder: Rsdr)
+        -> ReseedingRng<R, Rsdr>
+    {
         assert!(threshold <= ::core::i64::MAX as u64);
-        ReseedingRng {
-            rng: rng,
-            reseeder: reseeder,
-            threshold: threshold as i64,
-            bytes_until_reseed: threshold as i64,
+        let results_empty = R::Results::default();
+        ReseedingRng(
+            BlockRng {
+                core: Reseeder {
+                    core: rng,
+                    reseeder: reseeder,
+                    threshold: threshold as i64,
+                    bytes_until_reseed: threshold as i64,
+                },
+                index: results_empty.as_ref().len(), // generate on first use
+                results: results_empty,
+            }
+        )
+    }
+
+    /// Reseed the internal PRNG.
+    pub fn reseed(&mut self) -> Result<(), Error> {
+        self.0.core.reseed()
+    }
+}
+
+impl<R: BlockRngCore<u32> + SeedableRng, Rsdr: RngCore> RngCore for ReseedingRng<R, Rsdr> {
+    #[inline(always)]
+    fn next_u32(&mut self) -> u32 {
+        self.0.next_u32()
+    }
+
+    #[inline(always)]
+    fn next_u64(&mut self) -> u64 {
+        self.0.next_u64()
+    }
+
+    fn fill_bytes(&mut self, dest: &mut [u8]) {
+        self.0.fill_bytes(dest)
+    }
+
+    fn try_fill_bytes(&mut self, dest: &mut [u8]) -> Result<(), Error> {
+        self.0.try_fill_bytes(dest)
+    }
+}
+
+#[derive(Debug)]
+struct Reseeder<R, Rsdr> {
+    core: R,
+    reseeder: Rsdr,
+    threshold: i64,
+    bytes_until_reseed: i64,
+}
+
+impl<R, Rsdr> BlockRngCore<u32> for Reseeder<R, Rsdr>
+where R: BlockRngCore<u32> + SeedableRng,
+      Rsdr: RngCore
+{
+    type Results = <R as BlockRngCore<u32>>::Results;
+
+    fn generate(&mut self, results: &mut Self::Results) -> Result<(), Error> {
+        if self.bytes_until_reseed <= 0 {
+            // We want to reseed here, and generate results later in the
+            // function. If generating results fail, we should return the error
+            // from that. If generating results succeeded, but reseeding failed,
+            // we should return the error from reseeding.
+            // The only way to get this behaviour without destroying performance
+            // was to split part of the function out into a
+            // `reseed_and_generate` method.
```

Comment:
This was about branch prediction then, presumably? I can't find anything Rust-specific, and the answers [here](https://stackoverflow.com/questions/3702903/portable-branch-prediction-hints) contradict each other... so maybe this is the best option (and more portable than relying on which branch is most likely to be predicted).

Reply:
I don't think it had to do with branch prediction, but with the size of the function and LLVM trying to combine both `generate()` functions and both returns, moving things around and adding checks outside the single branch we have now.